composer/loggers/object_store_logger.py

```diff
@@ -0,0 +1,383 @@
+
+    def epoch_end(self, state: State, logger: Logger) -> None:
+        del state, logger  # unused
+        self._check_workers()
+
+    def _check_workers(self):
+        # Periodically check to see if any of the upload workers crashed
+        # They would crash if:
+        #   a) A file could not be uploaded, and the retry counter failed, or
+        #   b) allow_overwrite=False, but the file already exists,
+        for worker in self._workers:
+            if not worker.is_alive():
+                raise RuntimeError("Upload worker crashed. Please check the logs.")
+
+    def log_file_artifact(self, state: State, log_level: LogLevel, artifact_name: str, file_path: pathlib.Path, *,
+                          overwrite: bool):
+        if not self._should_log_artifact(state, log_level, artifact_name):
+            return
+        copied_path = os.path.join(self._upload_staging_folder, str(uuid.uuid4()))
+        copied_path_dirname = os.path.dirname(copied_path)
+        os.makedirs(copied_path_dirname, exist_ok=True)
+        shutil.copy2(file_path, copied_path)
+        object_name = self._format_object_name(artifact_name)
+        self._file_upload_queue.put_nowait((copied_path, object_name, overwrite))
+
+    def post_close(self):
+        # Cleaning up on post_close to ensure that all artifacts are uploaded
+        if self._finished is not None:
+            self._finished.set()
+        for worker in self._workers:
+            worker.join()
+        if self._tempdir is not None:
+            self._tempdir.cleanup()
+        self._workers.clear()
+
+    def get_uri_for_artifact(self, artifact_name: str) -> str:
+        """Get the object store provider uri for an artfact.
+
+        Args:
+            artifact_name (str): The name of an artifact.
+
+        Returns:
+            str: The uri corresponding to the uploaded location of the artifact.
+        """
+        obj_name = self._format_object_name(artifact_name)
+        provider_name = self._object_store_provider_hparams.provider
+        container = self._object_store_provider_hparams.container
+        provider_prefix = f"{provider_name}://{container}/"
+        return provider_prefix + obj_name.lstrip("/")
+
+    def _format_object_name(self, artifact_name: str):
+        """Format the ``artifact_name`` according to the ``object_name_format_string``."""
+        if self._run_name is None:
+            raise RuntimeError("The run name is not set. It should have been set on Event.INIT.")
+        key_name = self.object_name_format.format(
+            rank=dist.get_global_rank(),
+            local_rank=dist.get_local_rank(),
+            world_size=dist.get_world_size(),
+            local_world_size=dist.get_local_world_size(),
+            node_rank=dist.get_node_rank(),
+            artifact_name=artifact_name,
+            run_name=self._run_name,
+        )
+        key_name = key_name.lstrip('/')
+
+        return key_name
+
+
+def _validate_credentials(
```

Comment:
What happens if validation fails? Should document

Reply:
A libcloud error will be raised. Added this to the comments.